src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -74,121 +90,82 @@ private async Task AnalyzeProjectAsync(Project project, Document? specificDocume
+        private async Task<(Document document, DesignerAttributeData data)[]> ComputeLatestDataAsync(
+            Project project, Document? specificDocument, VersionStamp projectVersion, CancellationToken cancellationToken)
         {
-            using var storage = _storageService.GetStorage(solution);
-
-            foreach (var (doc, info, _) in latestInfos)
-            {
-                // Skip documents that didn't change contents/version at all.  No point in writing
-                // back out the exact same data as before.
-                if (info == null)
-                    continue;
-
-                using var memoryStream = new MemoryStream();
-                using var writer = new ObjectWriter(memoryStream);
-
-                PersistInfoTo(writer, info.Value, projectVersion);
-
-                memoryStream.Position = 0;
-                await storage.WriteStreamAsync(
-                    doc, DataKey, memoryStream, cancellationToken).ConfigureAwait(false);
-            }
-        }
-
-        private async Task<(Document, DesignerAttributeData? data, bool changed)[]> ComputeLatestInfosAsync(
-            Project project, VersionStamp projectVersion,
-            Document? specificDocument, CancellationToken cancellationToken)
-        {
-            using var storage = _storageService.GetStorage(project.Solution);
-
             var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
             var designerCategoryType = compilation.DesignerCategoryAttributeType();
 
-            using var _ = ArrayBuilder<Task<(Document, DesignerAttributeData?, bool changed)>>.GetInstance(out var tasks);
+            using var _ = ArrayBuilder<Task<(Document document, DesignerAttributeData data)>>.GetInstance(out var tasks);
             foreach (var document in project.Documents)
             {
                 // If we're only analyzing a specific document, then skip the rest.
                 if (specificDocument != null && document != specificDocument)
                     continue;
 
-                tasks.Add(ComputeDesignerAttributeDataAsync(
-                    storage, projectVersion, designerCategoryType, document, cancellationToken));
+                // If we don't have a path for this document, we cant proceed with it.
+                // We need that path to inform the project system which file we're referring to.
+                if (document.FilePath == null)
+                    continue;
+
+                // If nothing has changed at the top level between the last time we analyzed this document and now, then
+                // no need to analyze again.
+                if (_documentToLastReportedInformation.TryGetValue(document.Id, out var existingInfo) &&
+                    existingInfo.projectVersion == projectVersion)
```

Comment:
What's the threading guarantees here? Can RemoveDocumentAsync earlier in the file be running at the same time as this? Is there some risk that we'd remove the document, then process it here, and add it back?

Reply:
it cannot be running at the same time.  all the incremental-analyzer entrypoints are serialized.  the only concurrency is from me spitting out tons of tasks (that i when-all) in one of those incremental-analyzer entrypoints.

@mavasani to confirm.

Reply:
> all the incremental-analyzer entrypoints are serialized

Yes, that is correct